components/eamxx/src/physics/mam/eamxx_mam_aci_functions.hpp

```diff
@@ -0,0 +1,726 @@
+    const MAMAci::const_view_2d qi, const MAMAci::const_view_2d liqcldf,
+    const MAMAci::const_view_2d liqcldf_prev, const int icol, const int top_lev,
+    const int nlev,
+    // output
+    MAMAci::view_2d cloud_frac, MAMAci::view_2d cloud_frac_prev) {
+  //-------------------------------------------------------------
+  // Get old and new liquid cloud fractions when amount of cloud
+  // is above qsmall threshold value
+  //-------------------------------------------------------------
+  // cut-off for cloud amount (ice or liquid)
+  static constexpr auto qsmall = 1e-18;
+  Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, top_lev, nlev), KOKKOS_LAMBDA(int kk) {
+        if((qc(icol, kk) + qi(icol, kk)) > qsmall) {
+          cloud_frac(icol, kk)      = liqcldf(icol, kk);
+          cloud_frac_prev(icol, kk) = liqcldf_prev(icol, kk);
+        } else {
+          cloud_frac(icol, kk)      = 0;
+          cloud_frac_prev(icol, kk) = 0;
+        }
+      });
+}
+void store_liquid_cloud_fraction(
+    haero::ThreadTeamPolicy team_policy,
+    const mam_coupling::DryAtmosphere &dry_atmosphere,
+    const MAMAci::const_view_2d liqcldf,
+    const MAMAci::const_view_2d liqcldf_prev, const int top_lev, const int nlev,
+    // output
+    MAMAci::view_2d cloud_frac, MAMAci::view_2d cloud_frac_prev) {
+  MAMAci::const_view_2d qc = dry_atmosphere.qc;
+  MAMAci::const_view_2d qi = dry_atmosphere.qi;
+  Kokkos::parallel_for(
+      team_policy, KOKKOS_LAMBDA(const haero::ThreadTeam &team) {
+        const int icol = team.league_rank();
+        store_liquid_cloud_fraction(team, qc, qi, liqcldf, liqcldf_prev, icol,
+                                    top_lev, nlev,
+                                    // output
+                                    cloud_frac, cloud_frac_prev);
+      });
+}
+KOKKOS_INLINE_FUNCTION
+void compute_recipical_pseudo_density(const haero::ThreadTeam &team,
+                                      const MAMAci::const_view_2d pdel,
+                                      const int icol, const int nlev,
+                                      // output
+                                      MAMAci::view_2d rpdel) {
+  Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, 0u, nlev), KOKKOS_LAMBDA(int kk) {
+        EKAT_KERNEL_ASSERT_MSG(0 < pdel(icol, kk),
+                               "Error: pdel should be > 0.\n");
+        rpdel(icol, kk) = 1 / pdel(icol, kk);
+      });
+}
+void compute_recipical_pseudo_density(haero::ThreadTeamPolicy team_policy,
+                                      MAMAci::const_view_2d pdel,
+                                      const int nlev,
+                                      // output
+                                      MAMAci::view_2d rpdel) {
+  Kokkos::parallel_for(
+      team_policy, KOKKOS_LAMBDA(const haero::ThreadTeam &team) {
+        const int icol = team.league_rank();
+        compute_recipical_pseudo_density(team, pdel, icol, nlev,
```

Comment:
Maybe it's just me, but I see you have plenty of functions that have 2 versions, one to dispatch the outer loop, and one to be called from inside the loop. If the inner one is only used in the outer one, you could consider just using a lambda in the outer one, to keep the code smaller.

OTOH, dispatching lots of kernels is not great. If in all outer loops you just do
```
  const int icol = team.league_rank();
  inner_function_name(...);
```
you could consider merging all these kernels together, and doing all these calculations at once, if possible (it may not be, if they must happen at different times during the ACI step).

Reply:
Thanks. We will fix it in a follow-up PR (see issue #2858 )